File: NearestNeighboursAndClustering/src/labelDataset.cpp
#include "../headers/labelDataset.h"
#include "../headers/utilities.h"
#include <iostream>
#include <fstream>
#include <cstdio>
#include <cstring>
#include <cerrno>

LabelDataset::LabelDataset(std::string inputPath) {
    // Open labels binary file
    std::ifstream input(inputPath, std::ios::out | std::ios::binary);
    if(!input) {
        std::cout << "Cannot open file!:" << strerror(errno) << std::endl;
        this->valid = false;
        return;
    }

    // Read magic number
    input.read((char*)&this->magic_num,sizeof(this->magic_num));
    this->magic_num = SWAP_INT32(this->magic_num);

    // Read number of items
    input.read((char*)&this->num_of_items,sizeof(this->num_of_items));
    this->num_of_items = SWAP_INT32(this->num_of_items);

    // Read all labels
    uint8_t label;
    for (unsigned int i = 0;i < this->num_of_items;i++) {
        input.read((char*)&label,sizeof(label));
        this->labels.push_back(label);
    }

    // Close dataset binary file
    input.close();

    this->valid = true;
}

bool LabelDataset::isValid() {
    return this->valid;
}

uint8_t LabelDataset::getLabel(int labelIndex) {
    if (labelIndex < 0 || labelIndex >= this->labels.size()) {
        return -1;
    }
    return this->labels[labelIndex];
}

LabelDataset::~LabelDataset() {
    
}

File: autoencoder/labelDataset.py
import numpy as np

class LabelDataset:
    def __init__(self, file):
        dataset = open(file, "rb")
        self.classes = set()
        # Read header
        self.magic_num = int.from_bytes(dataset.read(4), byteorder='big', signed=False)
        self.num_of_items = int.from_bytes(dataset.read(4), byteorder='big', signed=False)
        # Read Images
        self.labels = []
        for _ in range(self.num_of_items):
            label = int.from_bytes(dataset.read(1), byteorder='big', signed=False)
            if label not in self.classes:
                self.classes.add(label)
            self.labels.append(label)
        
        dataset.close()

    def get_labels(self):
        return np.array(self.labels)

    def num_classes(self):
        return len(self.classes)